File: ozayn/desktop-app/api_client.py
import requests
import json
import os
import subprocess
import time

API_BASE = "http://127.0.0.1:8000/ozayn/backend/api"
ROOT_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__), "..", ".."))


class OzaynAPI:
    def __init__(self, base_url=API_BASE):
        self.base_url = base_url
        self.session_id = None
        self.user = None
        self.server_process = None
        self._ensure_server()

    def _ensure_server(self):
        try:
            r = requests.get(f"{self.base_url}/../..", timeout=2)
            return True
        except Exception:
            pass

        php_bin = "php"
        router = os.path.join(ROOT_DIR, "router.php")
        try:
            self.server_process = subprocess.Popen(
                [php_bin, "-S", "127.0.0.1:8000", "router.php"],
                cwd=ROOT_DIR,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
                preexec_fn=os.setsid
            )
            for _ in range(10):
                time.sleep(0.5)
                try:
                    requests.get(f"{self.base_url}/../..", timeout=1)
                    return True
                except Exception:
                    continue
        except Exception:
            pass
        return False

    def _request(self, method, endpoint, data=None):
        url = f"{self.base_url}{endpoint}"
        headers = {"Content-Type": "application/json"}
        if self.session_id:
            headers["X-Session-ID"] = self.session_id

        try:
            if method == "GET":
                r = requests.get(url, headers=headers, timeout=15)
            elif method == "POST":
                r = requests.post(url, headers=headers, json=data, timeout=15)
            elif method == "PUT":
                r = requests.put(url, headers=headers, json=data, timeout=15)
            elif method == "DELETE":
                r = requests.delete(url, headers=headers, timeout=15)
            else:
                return {"success": False, "error": "Invalid method"}

            return r.json()
        except requests.ConnectionError:
            return {"success": False, "error": "Cannot connect to server"}
        except Exception as e:
            return {"success": False, "error": str(e)}

    def login(self, username, password):
        result = self._request("POST", "/auth/login", {"username": username, "password": password})
        if result.get("success"):
            self.session_id = result["session_id"]
            self.user = result["user"]
        return result

    def register(self, username, password, email=None, full_name=None):
        data = {"username": username, "password": password}
        if email:
            data["email"] = email
        if full_name:
            data["full_name"] = full_name
        return self._request("POST", "/auth/register", data)

    def logout(self):
        self._request("POST", "/auth/logout")
        self.session_id = None
        self.user = None

    def send_chat(self, message, conversation_id=None, project_id=None):
        data = {"message": message}
        if conversation_id:
            data["conversation_id"] = conversation_id
        if project_id:
            data["project_id"] = project_id
        return self._request("POST", "/chat/send", data)

    def list_conversations(self):
        return self._request("GET", "/chat/list")

    def get_chat_history(self, conversation_id):
        return self._request("GET", f"/chat/history/{conversation_id}")

    def list_projects(self):
        return self._request("GET", "/projects/list")

    def create_project(self, name, description=""):
        return self._request("POST", "/projects/create", {"name": name, "description": description})

    def list_tasks(self):
        return self._request("GET", "/tasks/list")

    def create_task(self, title, description="", priority="medium", project_id=None):
        data = {"title": title, "description": description, "priority": priority}
        if project_id:
            data["project_id"] = project_id
        return self._request("POST", "/tasks/create", data)

    def update_task(self, task_id, status):
        return self._request("PUT", f"/tasks/{task_id}", {"status": status})

    def list_knowledge(self):
        return self._request("GET", "/knowledge/list")

    def add_knowledge(self, title, content, tags=None, project_id=None):
        data = {"title": title, "content": content, "tags": tags or []}
        if project_id:
            data["project_id"] = project_id
        return self._request("POST", "/knowledge/add", data)

    def cleanup(self):
        if self.server_process:
            try:
                self.server_process.terminate()
                self.server_process.wait(timeout=3)
            except Exception:
                try:
                    self.server_process.kill()
                except Exception:
                    pass
            self.server_process = None

File: ozayn/mobile-app/lib/api/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:shared_preferences/shared_preferences.dart';

class OzaynApiClient {
  static const String baseUrl = 'http://10.0.2.2:8000/ozayn/backend/api';

  String? sessionId;
  Map<String, dynamic>? user;

  Map<String, String> get _headers => {
    'Content-Type': 'application/json',
    if (sessionId != null) 'X-Session-ID': sessionId!,
  };

  Future<Map<String, dynamic>> _request(String method, String endpoint, {Map<String, dynamic>? data}) async {
    final uri = Uri.parse('$baseUrl$endpoint');
    try {
      http.Response response;
      if (method == 'GET') {
        response = await http.get(uri, headers: _headers).timeout(const Duration(seconds: 15));
      } else if (method == 'POST') {
        response = await http.post(uri, headers: _headers, body: jsonEncode(data ?? {})).timeout(const Duration(seconds: 15));
      } else if (method == 'PUT') {
        response = await http.put(uri, headers: _headers, body: jsonEncode(data ?? {})).timeout(const Duration(seconds: 15));
      } else if (method == 'DELETE') {
        response = await http.delete(uri, headers: _headers).timeout(const Duration(seconds: 15));
      } else {
        return {'success': false, 'error': 'Invalid method'};
      }
      return jsonDecode(response.body);
    } catch (e) {
      return {'success': false, 'error': e.toString()};
    }
  }

  // Auth
  Future<Map<String, dynamic>> login(String username, String password) async {
    final result = await _request('POST', '/auth/login', data: {'username': username, 'password': password});
    if (result['success'] == true) {
      sessionId = result['session_id'];
      user = result['user'];
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString('ozayn_session', sessionId!);
    }
    return result;
  }

  Future<Map<String, dynamic>> register(String username, String password, {String? email}) async {
    return await _request('POST', '/auth/register', data: {
      'username': username,
      'password': password,
      if (email != null) 'email': email,
    });
  }

  Future<bool> validateSession() async {
    try {
      final result = await _request('GET', '/chat/list');
      return result.containsKey('conversations');
    } catch (_) {
      return false;
    }
  }

  // Chat
  Future<Map<String, dynamic>> sendMessage(String message, {String? conversationId}) async {
    return await _request('POST', '/chat/send', data: {
      'message': message,
      if (conversationId != null) 'conversation_id': conversationId,
    });
  }

  Future<List<dynamic>> listConversations() async {
    final result = await _request('GET', '/chat/list');
    return result['conversations'] ?? [];
  }

  Future<List<dynamic>> getChatHistory(String conversationId) async {
    final result = await _request('GET', '/chat/history/$conversationId');
    return result['messages'] ?? [];
  }

  // Projects
  Future<List<dynamic>> listProjects() async {
    final result = await _request('GET', '/projects/list');
    return result['projects'] ?? [];
  }

  Future<Map<String, dynamic>> createProject(String name, {String? description}) async {
    return await _request('POST', '/projects/create', data: {
      'name': name,
      if (description != null) 'description': description,
    });
  }

  // Tasks
  Future<List<dynamic>> listTasks() async {
    final result = await _request('GET', '/tasks/list');
    return result['tasks'] ?? [];
  }

  Future<Map<String, dynamic>> createTask(String title, {String? priority}) async {
    return await _request('POST', '/tasks/create', data: {
      'title': title,
      if (priority != null) 'priority': priority,
    });
  }

  Future<Map<String, dynamic>> updateTask(dynamic taskId, String status) async {
    return await _request('PUT', '/tasks/$taskId', data: {'status': status});
  }

  // Knowledge
  Future<List<dynamic>> listKnowledge() async {
    final result = await _request('GET', '/knowledge/list');
    return result['results'] ?? [];
  }

  Future<Map<String, dynamic>> addKnowledge(String title, String content, {List<String>? tags}) async {
    return await _request('POST', '/knowledge/add', data: {
      'title': title,
      'content': content,
      if (tags != null) 'tags': tags,
    });
  }

  // ARWE
  Future<Map<String, dynamic>> arweStatus() async {
    return await _request('GET', '/arwe/status');
  }

  Future<Map<String, dynamic>> arweBriefing() async {
    return await _request('GET', '/arwe/briefing');
  }

  // Decisions
  Future<List<dynamic>> listDecisions() async {
    final result = await _request('GET', '/decisions/list');
    return result['decisions'] ?? [];
  }

  Future<Map<String, dynamic>> createDecision(String context) async {
    return await _request('POST', '/decisions/create', data: {'context': context});
  }

  // Audit
  Future<List<dynamic>> auditLog() async {
    final result = await _request('GET', '/audit/log');
    return result['log'] ?? [];
  }

  // System
  Future<Map<String, dynamic>> systemOverview() async {
    return await _request('GET', '/system/overview');
  }
}
